doc/developer/design/20210311_prometheus_source.md

```diff
@@ -0,0 +1,144 @@
+
+# Goals
+
+* Allow our users to read & make decisions based on our metrics data without setting up a whole metrics ingestion/presentation pipeline.
+
+* Let administrators inspect that metrics data (as much as metrics' attached data allows) together with the other available data we present in system tables.
+
+* Retain a long-enough history of metrics values to allow making decisions based on how the metrics values change (but short enough to not consume unbounded memory).
+
+# Non-Goals
+
+* Export our system table logs as prometheus (yet) -- this design covers only the "metrics import" direction.
+
+* Come up with a data retention scheme other than "older than n minutes gets dropped". If this was a full-featured time series database design, we'd want to thin metrics into minute/hour/day buckets. This isn't that.
+
+* (Maybe) don't even expose the `CREATE SOURCE` mechanics externally?
+
+# Description
+
+<!--
+    Describe the approach in detail. If there is no clear frontrunner, feel free to list all approaches in alternatives.
+    If applicable, be sure to call out any new testing/validation that will be required
+-->
+
+To import prometheus metrics, a process must regularly scrape an endpoint. Something like the following would be our external interface:
+
+```sql
+CREATE SOURCE metrics_table_name
+PROMETHEUS ENDPOINT 'http://localhost:9090/metrics'
+WITH (
+    metadata_table = "metrics_table_meta",
+    poll_interval = '1 minute'::interval,
+    retain_data = '5 minutes'::interval
+);
+```
+
+This creates two relations that have schemas similar to the following:
+
+```sql
+CREATE TABLE metrics_table_name (
+    metric TEXT,
+    ts TIMESTAMP,
+    labels JSONB,
+    value float8
+);
+
+CREATE TABLE metrics_table_meta (
+    metric TEXT,
+    metric_type TEXT,
+    help TEXT
+);
+```
+
+Once created, the source polls the endpoint once a minute (via the configured the `poll_interval` duration), and retains metric readings for 5 minutes.
+
+
+# Alternatives
+
+<!-- // Similar to the Description section. List of alternative approaches considered, pros/cons or why they were not chosen -->
+
+There are two major things that we could do differently to fulfill the brief, and a series of things that we could do to polish this design. I'll summarize the former under "Alternative approaches" and the latter under "Alternative details".
+
+## Alternative approaches
+
+### Create an internal metrics reporting channel instead (feeding metrics directly)
+
+We could add a mechanism for providing log events to the dataflow crate (or a crate that df and others depend on), where if a piece of code now increments a prometheus counter, it would instead send an event that gets captured by a listener on a channel, and increments the prometheus cuonter, as well as queue the event for updating of internal metrics views.
```

Comment:
This seems to to be about exposing internal Prometheus metrics directly into a system table. That's a different feature, right?

Reply:
What I want to do here that's the one non-negotiable requirement is getting our own prometheus metrics into system tables. That can happen with or without a prometheus scraper, so not really a different feature. Same end goal, different implementation mechanism.